_posts/2022-12-12-deltalake-spark-testing-learnings.md

```diff
@@ -0,0 +1,232 @@
+
+Key consideration here is to closely align the version of delta-spark and pyspark to Databricks as much as possible to try and prevent as much as possible scenarios where code may run in Databricks but fail in tests or vice versa. Few things you can do here are:
+
+1. Check out [databricks runtime release notes](https://docs.databricks.com/release-notes/runtime/releases.html) to ensure that the version you install is closely aligned with the version of Databricks you have running
+
+2. Check out [delta release notes](https://docs.delta.io/latest/releases.html) to ensure that delta-spark version you install is compatible with the Apache Spark version in step 1.
+
+
+### Configuring Docker For Local Test Execution
+
+Once we have the dependencies specificed in a requirements txt file or poetry toml file (if you use Poetry), the next step would be to create a docker compose solution for local development, so that it always installs a preset Python and OpenJDK versions and including pip dependencies, leaving less avenue for 'works on my machine' type of issues. This is relatively simple as you can see below:
+
+
+*Dockerfile*
+```
+FROM python:3.9.5
+
+RUN apt-get update && apt-get install openjdk-11-jdk -y
+RUN pip3 install -r requirements.txt
+
+CMD export PYSPARK_SUBMIT_ARGS='--packages io.delta:delta-core_2.12:2.2.0 pyspark-shell' \
+    pytest path/to/tests/**/*
+```
+
+If you are wondering why theres no Apache setup steps above, then remember that we are will be installing pyspark via pip instead which will suffice for local development activities. 
+
+Most of the other config is stock standard, with the exception of the `PYSPARK_SUBMIT_ARGS`. Its one of those things you need to set to run these tests as a workaround certain known issues with running locally. At the end of the day, these are just arguments that end up being supplied to `spark submit` which is a script that Spark uses to launch apps on a cluster. 
+
+Why you might still ask? well, the purpose behind each argument is as follows:
+
+1. io.delta:delta-core_2.12:2.1.0 - defines the maven artifact dependency of delta-core_2.21 for the version 2.1.0. You must match the version with precisely what you defined as the `delta-spark` pip dependency. If you are not sure about the artifact name, here is a reference: https://mvnrepository.com/artifact/io.delta/delta-core. The artifact name will be different but the version is whats important.
+
+2. pyspark-shell specifies to use the pyspark java class when running spark jobs locally. 
+
+
+*docker-compose.yml*
+```
+version: "3.9"
+
+services:
+  test:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    container_name: test
+    volumes:
+      - .:/src
+```
+
+Then we just need to run docker-compose up to re-run the tests whenever we make changes or create new ones locally. Of course, the volume mount in the compose file will depend on your project structure.
```

Comment:
```suggestion
Then we can run `docker-compose up` to re-run the tests whenever we make changes or create new ones locally. Of course, the volume mount in the compose file will depend on your project structure.
```

I try to avoid using words like `just` since it can come across as condesending - https://www.awesomecodereviews.com/respectful-code-review-feedback/